README.md
# Orpheus UI

## Overview

Orpheus UI is a comprehensive and easy-to-use UI component library built with modern technologies such as React, TypeScript, Radix and Tailwind CSS. 

A component library for React aiming to provide variety, polymorphism, and easy customization for developers needing to create modern interfaces.

Explore the live component previews: [Storybook](https://gilalvez.github.io/orpheus-ui/)

## Installation

At present, the package is **not** available for installation via npm packages as I am validating several use cases and conducting tests across different browsers and frameworks. After thorough feasibility testing, version **0.1.0** will be made available for installation.

## Technologies

- **React**: The foundational UI library for the project.
- **TypeScript**: Ensures static typing, increasing maintainability and developer efficiency.
- **Tailwind CSS**: Facilitates component styling, enabling customization and consistent design throughout the library.
- **Storybook**: Provides a live preview environment for the library's components, also serving as the main component documentation platform.
- **Radix-UI**: Assists in creating accessible design system components.
- **Vitest**: Handles testing to ensure high-quality components.
- **Parcel**: Helps in easy and efficient application bundling.

## Features

- Highly polymorphic, i.e., the rendered tag can be easily changed.
- Full TypeScript support for efficient development and maintainability.
- Customizable styles with Tailwind CSS, allowing for design consistency and flexibility.
- Allow to easily override styles by passing className.
- Most components are client components, but they offer many features uncommon in UI libraries, such as masking for text fields, multiple select with search, compound components, among others.
- Accessibility-focused components with the help of Radix-UI.
- Includes some logic utilities, such as hooks for implementing pagination and tables, augmenting functionality and facilitating development. 

## Peer Dependency

- TailwindCSS: Orpheus UI **dont** build any necessary CSS in bundle, its necessary Tailwind in project to use the components properly.

## Theming

Its possible extends/override Orpheus theme using the `withOrpheus` function in your `tailwind.config.js`:
```javascript
/** @type {import('tailwindcss').Config} */

const withOrpheus = require("orpheus-ui/dist/withOrpheus.js");

module.exports = withOrpheus({
  content: ["./src/**/*.{js,ts,jsx,tsx,mdx}"],
  theme: {
    extend: {
      colors: {
        primary: { // Easy color extends will reflect in all components
          50: "#fef2f3",
          100: "#fde6e7",
          200: "#fbd0d5",
          300: "#f7aab2",
          400: "#f27a8a",
          500: "#ea546c",
          600: "#d5294d",
          700: "#b31d3f",
          800: "#961b3c",
          900: "#811a39",
          950: "#48091a",
        },
        // secondary and tailwind color remains
        // its possible to override or add any color based in existing key
      },
    },
  },
  plugins: [],
});
```

The actual Orpheus preset is this:
```javascript
/** @type {import('tailwindcss').Config} */
module.exports = {
	darkMode: 'class',
	content: ['./node_modules/orpheus-ui/dist/index.js', './node_modules/orpheus-ui/dist/index.cjs'],
	theme: {
		screens: {
			sm: '640px',
			md: '768px',
			lg: '1024px',
			xl: '1280px',
			'2xl': '1536px',
			'3xl': '1920px',
		},
		colors: ({ colors }) => ({
			primary: {  
				50: '#eff8ff',
				100: '#dbeefe',
				200: '#c0e3fd',
				300: '#94d2fc',
				400: '#62b8f8',
				500: '#3d9af4',
				600: '#277de9',
				700: '#1f67d6',
				800: '#2053ad',
				900: '#1f4989',
				950: '#182d53',
			},
			secondary: {
				50: '#fef8ee',
				100: '#feefd6',
				200: '#fbdbad',
				300: '#f8c079',
				400: '#f49c43',
				500: '#f17f1e',
				600: '#e26414',
				700: '#bc4d12',
				800: '#953c17',
				900: '#783416',
				950: '#411809',
			},
			inherit: colors.inherit,
			current: colors.current,
			transparent: colors.transparent,
			black: colors.black,
			white: colors.white,
            // ... all others tailwind colors
		}),
		fontFamily: {
			sans: ['Roboto', 'sans-serif'],
			body: ['Roboto', 'sans-serif'],
			serif: ['Roboto Slab', 'Roboto Serif', 'Roboto', 'serif'],
			condensed: ['Roboto Condensed', 'Roboto', 'sans-serif'],
			mono: ['Roboto Mono', 'Roboto', 'monospace'],
		},
	},
	plugins: [],
};

```

## Components

Currently, the Storybook doesn't provide specific information about each component, but merely their variations and appearances. For now, I'll use this section to describe some possibilities of the existing components. Later, this information will be exclusively available in the Storybook.

#### Button

- Has  all tailwind color variations compound with design variations (filled, outlined and ghost) plus primary and secondary colors.
- asSlot prop: better informations here https://www.radix-ui.com/docs/primitives/utilities/slot

#### Text

- Text is a container component with 5 Types: Display(h1), Heading(h2), Title(h2), Label(label) and Body(p). Each type has 3 sizes: sm, md(default) and lg. Based in Material Text Theme of Flutter framework: [Flutter TextTheme](https://api.flutter.dev/flutter/material/TextTheme-class.html)
- Any text style is easily override by passing className. Including size.
- as prop: override base element tag
- Default usage:
```javascript
import { Text } from "orpheus-ui";

export default function MyComponent() {
  return (
    <>
      <Text.Display as='h2'>Hello</Text.Display>
      <Text.Heading size'lg'>Hello</Text.Heading>
      <Text.Title weight={700}>Hello</Text.Title>
    </>
  );
}
```

#### TextField

- For now TextField has many responsabilities: the container, label, input and error message.
- To override input styles use prop `classNames` that is a object with above elements. i.e. `classNames={{container: 'w-1/2', error: 'text-pink-500'}}`
- togglePasswordVisibility prop: Shows up a "show/hide password" icon
- Only has primary color usage, its possible to override but its many classes, you can change primary color in tailwind.config with `withOrpheus` feature.
- mask prop: built-in mask feature using `useMaskedInput` hook, highly based in [react-input-mask](https://github.com/sanniassin/react-input-mask), works with controlled and uncontrolled input


## Authors

- [@GilAlves](https://www.github.com/GilAlvez)



CHANGELOG.md
# Changelog

All notable changes to this library will be documented in this file.
The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/), and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [0.2.0] - 2023-05-28

### Added

- Project base configuration
- Documentation to explain the new components and utility.
- `Text` component for rendering text with predefined styles.
- `TextField` component for user text input with built-in mask.
- `Button` component for user interactions, with various style variants.
- `withOrpheus` utility to override Tailwind styles.

[Unreleased]: https://github.com/GilAlvez/orpheus-ui/compare/v0.2.0...develop
[0.2.0]: https://github.com/GilAlvez/orpheus-ui/releases/tag/v0.2.0


src/styles/theme.ts
export const screens = {
	xs: '(min-width: 412px)',
	sm: '(min-width: 640px)',
	md: '(min-width: 768px)',
	lg: '(min-width: 1024px)',
	xl: '(min-width: 1280px)',
	'2xl': '(min-width: 1536px)',
} as const;

export const colors = {
	black: '#000',
	white: '#fff',
	transparent: 'transparent',

	primary50: '#EFF8FF',
	primary100: '#DBEEFE',
	primary200: '#C0E3FD',
	primary300: '#94D2FC',
	primary400: '#62B8F8',
	primary500: '#3D9AF4',
	primary600: '#277CE9',
	primary700: '#1D5FC7',
	primary800: '#2053AD',
	primary900: '#1F4889',

	secondary50: '#FEF8EE',
	secondary100: '#FEEFD6',
	secondary200: '#FBDBAD',
	secondary300: '#F8C079',
	secondary400: '#F49C43',
	secondary500: '#F17F1E',
	secondary600: '#E26414',
	secondary700: '#BC4D12',
	secondary800: '#953C17',
	secondary900: '#783416',

	gray50: '#F9FAFB',
	gray100: '#F3F4F6',
	gray200: '#E5E7EB',
	gray300: '#D1D5DB',
	gray400: '#9CA3AF',
	gray500: '#6B7280',
	gray600: '#4B5563',
	gray700: '#374151',
	gray800: '#1F2937',
	gray900: '#111827',

	error50: '#FEF2F2',
	error100: '#FEE2E2',
	error200: '#FECACA',
	error300: '#FCA5A5',
	error400: '#F87171',
	error500: '#EF4444',
	error600: '#DC2626',
	error700: '#B91C1C',
	error800: '#991B1B',
	error900: '#7F1D1D',

	success50: '#F0FDF4',
	success100: '#DCFCE7',
	success200: '#BBF7D0',
	success300: '#86EFAC',
	success400: '#4ADE80',
	success500: '#22C55E',
	success600: '#16A34A',
	success700: '#15803D',
	success800: '#166534',
	success900: '#14532D',

	warning50: '#FEFCE8',
	warning100: '#FEF9C3',
	warning200: '#FEF08A',
	warning300: '#FDE047',
	warning400: '#FACC15',
	warning500: '#EAB308',
	warning600: '#CA8A04',
	warning700: '#A16207',
	warning800: '#854D0E',
	warning900: '#713F12',
} as const;

const negativeSpace = {
	'-px': '-1px',
	'-1': '-0.25rem',
	'-2': '-0.5rem',
	'-3': '-0.75rem',
	'-4': '-1rem',
	'-5': '-1.25rem',
	'-6': '-1.5rem',
	'-8': '-2rem',
	'-10': '-2.5rem',
	'-12': '-3rem',
	'-16': '-4rem',
	'-20': '-5rem',
	'-24': '-6rem',
	'-32': '-8rem',
	'-40': '-10rem',
	'-48': '-12rem',
	'-56': '-14rem',
	'-64': '-16rem',
	'-72': '18rem',
	'-80': '20rem',
	'-96': '24rem',
} as const;

export const space = {
	px: '1px',
	0: '0',
	1: '0.25rem',
	2: '0.5rem',
	3: '0.75rem',
	4: '1rem',
	5: '1.25rem',
	6: '1.5rem',
	7: '1.75rem',
	8: '2rem',
	9: '2.25rem',
	10: '2.5rem',
	11: '2.75rem',
	12: '3rem',
	14: '3.5rem',
	16: '4rem',
	20: '5rem',
	24: '6rem',
	32: '8rem',
	40: '10rem',
	48: '12rem',
	56: '14rem',
	64: '16rem',
	72: '18rem',
	80: '20rem',
	96: '24rem',
} as const;

export const sizes = { ...space };

export const fonts = {
	inter: 'Inter, sans-serif',
} as const;

export const fontSizes = {
	xs: '0.75rem',
	sm: '0.875rem',
	base: '1rem',
	lg: '1.125rem',
	xl: '1.25rem',
	'2xl': '1.5rem',
	'3xl': '1.875rem',
	'4xl': '2.25rem',
	'5xl': '3rem',
	'6xl': '3.75rem',
	'7xl': '4.5rem',
	'8xl': '6rem',
	'9xl': '8rem',
} as const;

export const fontWeights = {
	light: '300',
	normal: '400',
	medium: '500',
	semibold: '600',
	bold: '700',
} as const;

export const radii = {
	none: '0',
	xs: '0.125rem',
	sm: '0.25rem',
	md: '0.375rem',
	lg: '0.5rem',
	xl: '0.75rem',
	full: '9999px',
} as const;

export const borderWidths = {
	0: '0',
	1: '1px',
	2: '2px',
	4: '4px',
	8: '8px',
} as const;

export const shadows = {
	xs: '0 1px 2px 0 rgb(0 0 0 / 0.05)',
	sm: '0 1px 3px 0 rgb(0 0 0 / 0.1), 0 1px 2px -1px rgb(0 0 0 / 0.1)',
	md: '0 4px 6px -1px rgb(0 0 0 / 0.1), 0 2px 4px -2px rgb(0 0 0 / 0.1)',
	lg: '0 10px 15px -3px rgb(0 0 0 / 0.1), 0 4px 6px -4px rgb(0 0 0 / 0.1)',
	xl: '0 20px 25px -5px rgb(0 0 0 / 0.1), 0 8px 10px -6px rgb(0 0 0 / 0.1)',
	'2xl': '0 25px 50px -12px rgb(0 0 0 / 0.25)',
	inner: 'inset 0 2px 4px 0 rgb(0 0 0 / 0.05)',
	none: '0 0 #0000',
} as const;

export const lineHeights = {
	none: '1',
	tight: '1.25',
	snug: '1.375',
	normal: '1.5',
	relaxed: '1.625',
	loose: '2',
	3: '.75rem',
	4: '1rem',
	5: '1.25rem',
	6: '1.5rem',
	7: '1.75rem',
	8: '2rem',
	9: '2.25rem',
	10: '2.5rem',
	11: '2.75rem',
	12: '3rem',
} as const;

export const zIndices = {
	auto: 'auto',
	0: '0',
	10: '10',
	20: '20',
	30: '30',
	40: '40',
	50: '50',
} as const;

// ------------ EXTENDS ------------ //

export const extendMargin = {
	auto: 'auto',
	...space,
	...negativeSpace,
};
